File: src/Python/day_02.py
from aoc import get_lines
from itertools import accumulate
from operator import mul
from typing import Tuple, List


def parse_input(lines: List[str]) -> List[Tuple[str, int]]:
    return list(map(lambda x: (x[0], int(x[1])), map(lambda x: x.split(), lines)))


def part_1(lines: List[Tuple[str, int]]) -> int:
    x = sum(value for token, value in lines if token == "forward")
    depth = sum(value for token, value in lines if token == "down") - sum(
                value for token, value in lines if token == "up")
    return x * depth


def part_2(lines: List[Tuple[str, int]]) -> int:
    aims = map(lambda x: 0 if x[0] == 'forward' else (-x[1] if x[0] == 'up' else x[1]), lines)
    forwards = list(map(lambda x: x[1] if x[0] == 'forward' else 0, lines))
    depths = map(mul, forwards, accumulate(aims))
    return sum(depths) * sum(forwards)


def main():
    lines = get_lines("input_02.txt")
    lines = parse_input(lines)
    print("Part 1 :", part_1(lines))
    print("Part 2 :", part_2(lines))


if __name__ == '__main__':
    main()

File: src/Haskell/day_02.hs
{-# LANGUAGE MultiWayIf #-}

data Command = Up Int | Down Int | Forward Int deriving Show

commval :: Command -> Int
commval (Up i) = i 
commval (Down i) = i 
commval (Forward i) = i 
f (Forward _ ) = True
f _ = False
u (Up _) = True
u _ = False
d (Down _) = True
d _ = False

main = do
    contents <- readFile "../../inputs/input_02.txt"
    let comm =  map parseLine . lines $ contents
    print $ part1 $ comm
    print $ part2 (depths (forwards comm) (scanl (+) 0 (aims comm) )) $ forwards comm
 
aims :: [Command] -> [Int]
aims comm =  map (\x -> if 
                        | f(x) -> 0 
                        | otherwise -> commval x) comm

forwards :: [Command] -> [Int]
forwards comm =  map (\x -> if 
                        | f(x) -> commval x 
                        | otherwise -> 0) comm

depths :: [Int] -> [Int] -> [Int]
depths x y = zipWith (*)  x y

part1 :: [Command] -> Int
part1 comms = sum [commval(x) | x <- comms, f(x)]  * 
                (sum  [commval(x) | x <- comms, u(x)] + 
                 sum  [commval(x) | x <- comms, d(x)]) 

part2 :: [Int] -> [Int]-> Int
part2 x y = sum (x) * sum (y) 

parseLine :: String -> Command
parseLine = x . words 
    where 
    x ["forward", num] = Forward (read num)
    x ["up", num] = Up (- read num)
    x ["down", num] = Down (read num)





